src/libmongoc/src/mongoc/mongoc-server-monitor.c

```diff
@@ -0,0 +1,1241 @@
+ * Caller must hold topology lock.
+ */
+mongoc_server_monitor_t *
+mongoc_server_monitor_new (mongoc_topology_t *topology,
+                           mongoc_server_description_t *init_description)
+{
+   mongoc_server_monitor_t *server_monitor;
+
+   server_monitor = bson_malloc0 (sizeof (*server_monitor));
+   server_monitor->description =
+      mongoc_server_description_new_copy (init_description);
+   server_monitor->server_id = init_description->id;
+   server_monitor->topology = topology;
+   server_monitor->heartbeat_frequency_ms =
+      topology->description.heartbeat_msec;
+   server_monitor->min_heartbeat_frequency_ms =
+      topology->min_heartbeat_frequency_msec;
+   server_monitor->connect_timeout_ms = topology->connect_timeout_msec;
+   server_monitor->uri = mongoc_uri_copy (topology->uri);
+/* TODO CDRIVER-3682: Do not retrieve ssl opts from topology scanner. They
+ * should be stored somewhere else. */
+#ifdef MONGOC_ENABLE_SSL
+   if (topology->scanner->ssl_opts) {
+      server_monitor->ssl_opts = bson_malloc0 (sizeof (mongoc_ssl_opt_t));
+
+      _mongoc_ssl_opts_copy_to (
+         topology->scanner->ssl_opts, server_monitor->ssl_opts, true);
+   }
+#endif
+   memcpy (&server_monitor->apm_callbacks,
+           &topology->description.apm_callbacks,
+           sizeof (mongoc_apm_callbacks_t));
+   server_monitor->apm_context = topology->description.apm_context;
+   server_monitor->initiator = topology->scanner->initiator;
+   server_monitor->initiator_context = topology->scanner->initiator_context;
+   mongoc_cond_init (&server_monitor->shared.cond);
+   bson_mutex_init (&server_monitor->shared.mutex);
+   return server_monitor;
+}
+
+/* Creates a stream and performs the initial ismaster handshake.
+ *
+ * Called only by server monitor thread.
+ * Returns true if both connection and handshake succeeds.
+ * Returns false and sets error otherwise.
+ * ismaster_reply is always initialized.
+ */
+static bool
+_server_monitor_setup_connection (mongoc_server_monitor_t *server_monitor,
+                                  bson_t *ismaster_reply,
+                                  int64_t *start_us,
+                                  bson_error_t *error)
+{
+   bson_t cmd = BSON_INITIALIZER;
+   const bson_t *handshake;
+   bool ret = false;
+
+   ENTRY;
+
+   BSON_ASSERT (!server_monitor->stream);
+   bson_init (ismaster_reply);
+
+   server_monitor->more_to_come = false;
+
+   /* Using an initiator isn't really necessary. Users can't set them on
+    * pools. But it is used for tests. */
+   if (server_monitor->initiator) {
+      server_monitor->stream =
+         server_monitor->initiator (server_monitor->uri,
+                                    &server_monitor->description->host,
+                                    server_monitor->initiator_context,
+                                    error);
+   } else {
+      void *ssl_opts_void = NULL;
+
+#ifdef MONGOC_ENABLE_SSL
+      ssl_opts_void = server_monitor->ssl_opts;
```

Comment:
Does this need an explicit case, or is that redundant since `ssl_opts_void` is already a `void *`?